static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js

```diff
@@ -302,40 +305,43 @@ const buildPageTargetting = (
     return pageTargeting;
 };
 
-const getPageTargeting = (): { [key: string]: mixed } => {
-    if (Object.keys(myPageTargetting).length !== 0) return myPageTargetting;
-
-    onConsentChange(state => {
-        let canRun: boolean | null;
-        if (state.ccpa) {
-            // CCPA mode
-            canRun = !state.ccpa.doNotSell;
-        } else if (state.tcfv2) {
-            // TCFv2 mode
-            canRun = state.tcfv2.consents
-                ? Object.keys(state.tcfv2.consents).length > 0 &&
-                  Object.values(state.tcfv2.consents).every(Boolean)
-                : false;
-        } else if (state.aus) {
-            // AUS mode
-            canRun = state.aus.personalisedAdvertising;
-        } else canRun = false;
-
-        if (canRun !== latestConsentCanRun) {
-            const ccpaState = state.ccpa ? state.ccpa.doNotSell : null;
-            const eventStatus = state.tcfv2 ? state.tcfv2.eventStatus : 'na';
-            myPageTargetting = buildPageTargetting(
-                canRun,
-                ccpaState,
-                eventStatus
-            );
-            latestConsentCanRun = canRun;
-        }
+const getPageTargeting = async (): Promise<PageTargettingLoose> =>
+    new Promise(resolve => {
+        if (Object.keys(myPageTargetting).length !== 0)
+            return resolve(myPageTargetting);
+
+        onConsentChange(async state => {
+            let canRun: boolean | null;
+            if (state.ccpa) {
+                // CCPA mode
+                canRun = !state.ccpa.doNotSell;
+            } else if (state.tcfv2) {
+                // TCFv2 mode
+                canRun = state.tcfv2.consents
+                    ? Object.keys(state.tcfv2.consents).length > 0 &&
+                      Object.values(state.tcfv2.consents).every(Boolean)
+                    : false;
+            } else if (state.aus) {
+                // AUS mode
+                canRun = state.aus.personalisedAdvertising;
+            } else canRun = false;
+
+            if (canRun !== latestConsentCanRun) {
+                const ccpaState = state.ccpa ? state.ccpa.doNotSell : null;
+                const eventStatus = state.tcfv2
+                    ? state.tcfv2.eventStatus
+                    : 'na';
+                myPageTargetting = await buildPageTargetting(
+                    canRun,
+                    ccpaState,
+                    eventStatus
+                );
+                latestConsentCanRun = canRun;
+                resolve(myPageTargetting);
```

Comment:
This may work, but it feels really wierd to me. This is a callback that's fired in three ways:
1 - We're waiting for CMP to initalize, it'll be fired when it's done.
2 - CMP already initalized, fired immediately with current state.
3 - Consent has changed (?) So every time someone goes into the privacy manager?

In case 3, where it's already fired once, what's the meaning of 'resolve'. You can only resolve a promise once and it was resolved when the CMP initalised the first time.

Discuss on chat when you see this.

Reply:
will `buildPageTargetting` return different results if consent state changes? if so then yeah resolving once here isn't enough